### PR TITLE
Change witnesses to set

### PR DESCRIPTION
**Pull request by aljones15**
- [x] clean up jsdoc
- [x] remove witnessSet
- [x] grep for all Array only methods
- [x] change all `includes` to `has`
- [x] change `some` to `has`
- [x] remove all `witness =` that are not defaults in destructuring
- [x] change all defaults in destructuring to `new Set()`
- [x] remove previous witness sets
- [x] ensure no set is used with a mongodb query
- [x] change test data to use `Set` over `Array`
- [x] replace all instances of `length` with `size`
- [x] All tests passing

Gist with test results before and after changes and also relevant code snippets before and after:
https://gist.github.com/aljones15/68b64c86c23ec9d5fffe0473583db193

### CODE REVIEW

File: lib/continuity.js [hunk 1]
```diff
@@ -184,7 +184,6 @@ function _getTails({history, witnesses, state = null}) {
   }
 
   const witnessTails = new Map();
-  const witnessSet = new Set(witnesses);
```
Comment (gannan08) [inline lib/continuity.js:187]: Here we instantiate a new `Set` for use within the function. The code was changed to use the set passed in as a parameter. Are we modifying the `witnesses` parameter being passed in? Does the function caller expect the same variable to not be modified?

File: lib/worker/merge.js [hunk 2]
```diff
@@ -20,25 +20,24 @@ api.merge = async ({
   worker, priorityPeers = [], witnesses = worker.consensusState.witnesses,
   basisBlockHeight, peerWitnessParentThreshold
 }) => {
-  if(witnesses.length === 0) {
+  if(witnesses.size === 0) {
     throw new TypeError('"witnesses" must be a non-empty array.');
   }
-  const witnessSet = new Set(witnesses);
```
Comment (gannan08) [inline lib/worker/merge.js:26]: Same concern as previous comment around preserving the mutability of parameters.

File: test/mocha/helpers.js [hunk 3]
```diff
@@ -166,6 +166,7 @@ api.buildHistory = async ({historyId, mockData, nodes, witnesses} = {}) => {
   // set workers to use passed witnesses when merging
   for(const key in nodes) {
     const ledgerNode = nodes[key];
+    // double check to ensure witnesses is a Set
```
Comment (gannan08) [inline test/mocha/helpers.js:169]: Is this comment relevant?

File: lib/blocks.js [hunk 4]
```diff
@@ -237,7 +237,7 @@ async function _setRequiredBlockHeight({ledgerNode, witnesses, blockHeight}) {
   // FIXME: make this a covered query
   const {collection} = ledgerNode.storage.events;
   const result = await collection.updateMany({
-    'meta.continuity2017.creator': {$in: witnesses},
+    'meta.continuity2017.creator': {$in: Array.from(witnesses)},
```
Comment (gannan08) [inline lib/blocks.js:240]: It seems awkward for a storage API to convert a Set to an Array? My preference would be to revert this change. I understand the convenience of using a Set instead, so this is not a required fix.

**Review by gannan08 — COMMENTED** [review]
No approval from me yet. Have a few comments.

File: lib/worker/merge.js (hunk 2 above)
Comment (aljones15) [reply lib/worker/merge.js:26]: I'm not showing any mutations occurring on the `Set` itself in this file:

```
andrewjones@Andrews-MacBook-Pro-2 bedrock-ledger-consensus-continuity % grep -inr 'delete(' ./lib/worker/merge.js 
andrewjones@Andrews-MacBook-Pro-2 bedrock-ledger-consensus-continuity % grep -inr 'clear(' ./lib/worker/merge.js
andrewjones@Andrews-MacBook-Pro-2 bedrock-ledger-consensus-continuity % grep -inr 'add(' ./lib/worker/merge.js
./lib/worker/merge.js:324:  parentHashes.add(head.eventHash);
./lib/worker/merge.js:338:      parentHashes.add(peerHead.eventHash);
./lib/worker/merge.js:339:      peerCreatorSet.add(peerHead.creator);
./lib/worker/merge.js:347:    parentHashes.add(peerCommitmentParent.eventHash);
./lib/worker/merge.js:348:    peerCreatorSet.add(peerCommitmentParent.creator);
./lib/worker/merge.js:357:      parentHashes.add(eventHash);
./lib/worker/merge.js:391:      parentHashes.add(committedTo.eventHash);
./lib/worker/merge.js:392:      peerCreatorSet.add(committedTo.creator);
./lib/worker/merge.js:499:        pendingLocalRegularEventHashes.add(eventHash);
./lib/worker/merge.js:500:        parentHashes.add(eventHash);
./lib/worker/merge.js:516:        parentHashes.add(peerHead.eventHash);
./lib/worker/merge.js:517:        peerCreatorSet.add(peerHead.creator);
```

File: lib/continuity.js (hunk 1 above)
Comment (aljones15) [reply lib/continuity.js:187]: No `Set` based mutations seem to be occurring on `witnesses` in this file:

```
andrewjones@Andrews-MacBook-Pro-2 bedrock-ledger-consensus-continuity % grep -inr 'delete(' ./lib/continuity.js  
andrewjones@Andrews-MacBook-Pro-2 bedrock-ledger-consensus-continuity % grep -inr 'clear(' ./lib/continuity.js
andrewjones@Andrews-MacBook-Pro-2 bedrock-ledger-consensus-continuity % grep -inr 'add(' ./lib/continuity.js
./lib/continuity.js:126:    creators.add(_getCreator(e));
./lib/continuity.js:315:      mergeEventHashes.add(event.eventHash);
./lib/continuity.js:338:        parentHashes.add(parentHash);
./lib/continuity.js:982:      fPlusOneSupport.add(tallyResult);
./lib/continuity.js:1043:        union.add(e);
./lib/continuity.js:1056:          union.add(e);
./lib/continuity.js:1061:        union.add(event);
./lib/continuity.js:1108:  supportSets.add(union);
./lib/continuity.js:1247:        endorsers.add(witness);
./lib/continuity.js:1248:        _endorsers.add(witness);
./lib/continuity.js:1369:        priorityPeers.add(witness);
./lib/continuity.js:1381:        priorityPeers.add(witness);
./lib/continuity.js:1388:        priorityPeers.add(witness);
./lib/continuity.js:1397:        priorityPeers.add(witness);
./lib/continuity.js:1406:        priorityPeers.add(witness);
./lib/continuity.js:1419:        proposalEndorsers.add(otherWitness);
./lib/continuity.js:1425:      priorityPeers.add(witness);
./lib/continuity.js:1434:      priorityPeers.add(witness);
./lib/continuity.js:1602:          sortedSet.add(parent);
./lib/continuity.js:1605:      sortedSet.add(event);
./lib/continuity.js:1606:      round.add(event);
./lib/continuity.js:1619:    sortedHashes.add(eventHash);
```

File: test/mocha/helpers.js (hunk 3 above)
Comment (aljones15) [reply test/mocha/helpers.js]: this is definitely my bad and I have removed this comment here: https://github.com/digitalbazaar/bedrock-ledger-consensus-continuity/pull/248/commits/e2b620c0750bebb9164eaa56f10625a59e85e2f2

File: lib/blocks.js (hunk 4 above)
Comment (dlongley) [reply lib/blocks.js:240]: It should just be an array in the database, this item about using sets was supposed to be a minimal internal API change.

Comment (dlongley) [reply lib/blocks.js:240]: Yeah, so this function will be moving to the database module and it should take an array for `witnesses`. Any `Set` stuff should be internal to this module.

File: lib/continuity.js (hunk 1 above)
Comment (dlongley) [reply lib/continuity.js:187]: If there are no mutations as suggested above, then this change LGTM.

File: lib/continuity.js [hunk 5]
```diff
@@ -1306,8 +1305,8 @@ function _getPriorityPeers({
 
   If creating a new merge event would be helpful, prioritize the witness.
   */
-  const supermajority = api.supermajority(witnesses.length);
-  const f = api.maximumFailures(witnesses.length);
+  const supermajority = api.supermajority(witnesses.size);
+  const f = api.maximumFailures(witnesses.size);
```
Comment (dlongley) [inline lib/continuity.js:1309]: The changes from here on up LGTM so far.

File: lib/events.js [hunk 6]
```diff
@@ -663,7 +663,7 @@ async function _createLocalEventRecord({worker, event, eventHash}) {
   // merge event that merges this event reaches consensus or if the peer
   // becomes a witness on a subsequent block
   const {blockHeight, witnesses} = consensusState;
-  const requiredBlockHeight = witnesses.includes(localPeerId) ?
+  const requiredBlockHeight = witnesses.has(localPeerId) ?
```
Comment (dlongley) [inline lib/events.js:666]: This LGTM, it makes sense for `witnesses` to be a `Set` inside `consensusState`.

File: lib/history.js [hunk 7]
```diff
@@ -440,7 +440,7 @@ async function _getAcceptableHistory({
         ({witnesses} = await _witnesses.getBlockWitnesses(
           {ledgerNode, blockHeight: localBlockHeight + 1}));
       }
-      if(witnesses.includes(creator)) {
+      if(witnesses.has(creator)) {
```
Comment (dlongley) [inline lib/history.js:443]: I think it's ok to change `_witnesses.getBlockWitnesses` to return a `Set` of witnesses, +1.

File: lib/peerEvents.js [hunk 8]
```diff
@@ -1167,7 +1167,7 @@ async function _checkPeerCommitments({
   const peerIds = [];
   for(const eventRecord of mergeEvents) {
     const {meta: {continuity2017: {creator, generation}}} = eventRecord;
-    if(generation === 1 && !witnesses.includes(creator)) {
+    if(generation === 1 && !witnesses.has(creator)) {
```
Comment (dlongley) [inline lib/peerEvents.js:1170]: The changes in this file LGTM.

File: lib/peers.js [hunk 9]
```diff
@@ -369,12 +379,12 @@ function _samplePeers({peers, start, end = peers.length - 1}) {
...
 // it to submit its own merge events
 async function _isPeerCommitmentRequired({
-  ledgerNode, peerId, localPeerId, witnesses = []
+  ledgerNode, peerId, localPeerId, witnesses = new Set()
 } = {}) {
   // if peer is presently a witness, then no commitment is required; note that
   // this only applies *at the block height* for the given witnesses, so only
   // pass `witnesses` with care
-  if(witnesses.includes(peerId)) {
+  if(witnesses.has(peerId)) {
```
Comment (dlongley) [inline lib/peers.js:387]: I *think* the changes here are ok.

File: lib/worker/GossipPeer.js [hunk 10]
```diff
@@ -142,7 +142,7 @@ module.exports = class GossipPeer {
       _peer,
       worker: {ledgerNode, consensusState: {blockHeight, witnesses}}
     } = this;
-    const isWitness = witnesses.some(id => id === _peer.id);
+    const isWitness = witnesses.has(_peer.id);
```
Comment (dlongley) [inline lib/worker/GossipPeer.js:145]: These changes are good since witnesses comes from the `consensusState`.

File: lib/worker/Worker.js [hunk 11]
```diff
@@ -412,15 +412,15 @@ module.exports = class Worker {
...
+  _selectWithheld({witnesses} = {}) {
     const {withheldCache} = this;
     if(withheldCache.itemCount === 0) {
       return null;
     }
     const keys = withheldCache.keys();
     for(const creator of keys) {
       // skip any current witnesses
-      if(witnessSet.has(creator)) {
+      if(witnesses.has(creator)) {
```
Comment (dlongley) [inline lib/worker/Worker.js:423]: These changes LGTM.

File: lib/worker/merge.js [hunk 12]
```diff
@@ -20,25 +20,24 @@ api.merge = async ({
   worker, priorityPeers = [], witnesses = worker.consensusState.witnesses,
   basisBlockHeight, peerWitnessParentThreshold
 }) => {
-  if(witnesses.length === 0) {
+  if(witnesses.size === 0) {
     throw new TypeError('"witnesses" must be a non-empty array.');
```
Comment (dlongley) [inline lib/worker/merge.js:24]:
```suggestion
    throw new TypeError('"witnesses" must be a non-empty set.');
```

Comment (dlongley) [reply lib/worker/merge.js:24]: I didn't look that closely to see if other TypeErrors need to be updated, please take a quick look.

File: lib/worker/merge.js [hunk 13]
```diff
@@ -512,7 +511,7 @@ async function _prepareNextMerge({
   // speed along consensus
   if(isWitness && remainingSlots > 0) {
     for(const peerHead of remainingChildlessPeerHeads) {
-      if(witnessSet.has(peerHead.creator) &&
+      if(witnesses.has(peerHead.creator) &&
```
Comment (dlongley) [inline lib/worker/merge.js:514]: Changes LGTM here, too.

**Review by dlongley — CHANGES_REQUESTED** [review]
Ok, I think the only red flag was the changes to the code that will be moving to the database layer. Once fixed up I think it will be ready for approval. Thanks!

File: lib/blocks.js (hunk 4 above)
Comment (aljones15) [reply lib/blocks.js]: this was addressed here: https://github.com/digitalbazaar/bedrock-ledger-consensus-continuity/pull/248/commits/3512f5115a4a2649b4bf07b905e396332f02e271

please note that:

```
andrewjones@Andrews-MacBook-Pro-2 bedrock-ledger-consensus-continuity % grep -inr _setRequiredBlockHeight ./lib 
./lib/blocks.js:154:    _setRequiredBlockHeight(
./lib/blocks.js:247:async function _setRequiredBlockHeight({ledgerNode, witnesses, blockHeight}) {
./lib/peerEvents.js:1043:      _setRequiredBlockHeight(
./lib/peerEvents.js:1060:        _setRequiredBlockHeight(
./lib/peerEvents.js:1129:    _setRequiredBlockHeight({eventRecord, blockHeight, eventMap});
./lib/peerEvents.js:1233:function _setRequiredBlockHeight({eventRecord, blockHeight, eventMap}) {
```

`_setRequiredBlockHeight` in `./lib/peerEvents.js` is a different function than `_setRequiredBlockHeight` in `./lib/blocks.js`.

**Review by dlongley — APPROVED** [review]
Thanks!
